Issue1921 organic rankine cycle

hcasperfu opened the pull request:
This closes #1921.

@FWuellhorst -
Greetings! Do you mind reviewing this PR?
A (hopefully) comprehensive documentation of this model is in the info tab of `IBPSA.Fluid.CHPs.OrganicRankine.Cycle`.
We are hoping to have this model clear IBPSA repo before we present it in the 2024 American Modelica Conference on 14-16 Oct.

FWuellhorst commented:
@hcasperfu I should be able to provide a review in the next two weeks, if that fits with your schedule :)

hcasperfu commented:
@FWuellhorst -
That would be excellent! Thanks!

FWuellhorst reviewed (changes requested):
Overall, the implementation and model looks good, with a thourugh documentation and nice plots. 
Aside from my in-line comments, I have some questions / concerns:

1. How can users provide a constant value for evaporation temperature? Is the use-case always a constant temperature waste-heat? At least for Carnot-Batteries, the inlet temperature will decrease over time and thus the TEva would need to decrease, as well. Compared to heat pumps, I would expect the internal control to adjust the evaporating pressure / temperature in such a way that the expander outlet is outside "the dome". Or does it all just work based on the amount of mass flow rate in ORC? In the end, it would be nice to see some comparison to data provided in a ORC datasheet or similar to see if nominal evaporating temperatures, mass flow rates, and pinch points are typically known or not. 
2. As no actual validation is given, I can not review if the model itself makes sense or not. The equations and physics look good, but whether the constant pinch points and evaporating temperature is realistic is hard to say.
3. Wouldn't have expected anything different, but most naming and docs look good. The script we have in AixLib to adhere to naming guidelines found the following possibly relevant violations. The specific names (q, w) and endings (reg, T, set, actual) are not listed in the Buildings guideline as far as I know. 

```
IBPSA\Fluid\CHPs\OrganicRankine\Examples\ORCHotWater.mo
3: Name 'TWatOut_set' contains parts with more/less than 3 characters or which are not part of special cases. Affected parts: Out_set. Affected line: Modelica.Blocks.Sources.Constant TWatOut_set(k=55 + 273.15) "Set point of hot water output" annotation (Placement(transformation(extent={{0,0},{20,20}})));

4: Missing documentation. Affected line: Modelica.Blocks.Logical.And and1 annotation (Placement(transformation(extent={{-140,40},{-120,60}})));

5: Missing documentation. Affected line: IBPSA.Fluid.Sensors.TemperatureTwoPort senTColOut( redeclare final package Medium = MediumCol, m_flow_nominal=mCol_flow_nominal, T_start=TCol_start) annotation (Placement( transformation( extent={{10,-10},{-10,10}}, rotation=0, origin={-70,-40})));



IBPSA\Fluid\CHPs\OrganicRankine\BaseClasses\InterpolateStates.mo
1: Name 'qCon' contains parts with more/less than 3 characters or which are not part of special cases. Affected parts: q. Affected line: Modelica.Units.SI.SpecificEnergy qCon = hExpOut - hPumInl "Condenser specific energy transfer out of the cycle";

2: Name 'wExp' contains parts with more/less than 3 characters or which are not part of special cases. Affected parts: w. Affected line: Modelica.Units.SI.SpecificEnergy wExp = hExpOut - hExpInl "Expander specific work";

3: Name 'wPum' contains parts with more/less than 3 characters or which are not part of special cases. Affected parts: w. Affected line: Modelica.Units.SI.SpecificEnergy wPum = (pEva - pCon) / (rhoLiq * etaPum) "Pump specific work";

4: Name 'h_reg' contains parts with more/less than 3 characters or which are not part of special cases. Affected parts: h_reg. Affected line: protected Modelica.Units.SI.SpecificEnthalpy h_reg = hExpOutDry - hSatVapCon "For regularisation; if > 0, dry cycle";

5: Name 'rhoLiqDer_T' contains parts with more/less than 3 characters or which are not part of special cases. Affected parts: Der_. Affected line: final parameter Real rhoLiqDer_T[pro.n]= IBPSA.Utilities.Math.Functions.splineDerivatives( x = pro.T, y = pro.rhoLiq, ensureMonotonicity = true) "Derivative of saturated liquid density vs. temperature for cubic spline";

6: Name 'sSatLiqDer_T' contains parts with more/less than 3 characters or which are not part of special cases. Affected parts: Der_. Affected line: final parameter Real sSatLiqDer_T[pro.n]= IBPSA.Utilities.Math.Functions.splineDerivatives( x = pro.T, y = pro.sSatLiq, ensureMonotonicity = true) "Derivative of saturated liquid entropy vs. temperature for cubic spline";

7: Name 'sSatVapDer_T' contains parts with more/less than 3 characters or which are not part of special cases. Affected parts: Der_. Affected line: final parameter Real sSatVapDer_T[pro.n]= IBPSA.Utilities.Math.Functions.splineDerivatives( x = pro.T, y = pro.sSatVap, ensureMonotonicity = false) "Derivative of saturated vapor entropy vs. temperature for cubic spline";

8: Name 'hSatLiqDer_T' contains parts with more/less than 3 characters or which are not part of special cases. Affected parts: Der_. Affected line: final parameter Real hSatLiqDer_T[pro.n]= IBPSA.Utilities.Math.Functions.splineDerivatives( x = pro.T, y = pro.hSatLiq, ensureMonotonicity = true) "Derivative of saturated liquid enthalpy vs. temperature for cubic spline";

9: Name 'hSatVapDer_T' contains parts with more/less than 3 characters or which are not part of special cases. Affected parts: Der_. Affected line: final parameter Real hSatVapDer_T[pro.n]= IBPSA.Utilities.Math.Functions.splineDerivatives( x = pro.T, y = pro.hSatVap, ensureMonotonicity = false) "Derivative of saturated vapor enthalpy vs. temperature for cubic spline";

IBPSA\Fluid\CHPs\OrganicRankine\Cycle.mo
1: Name 'dTPinEva_set' contains parts with more/less than 3 characters or which are not part of special cases. Affected parts: Eva_set. Affected line: parameter Modelica.Units.SI.TemperatureDifference dTPinEva_set( final min = 0) = 5 "Set evaporator pinch point temperature difference" annotation(Dialog(group="Evaporator"));

6: Name 'mWor_flow_hysteresis' contains parts with more/less than 3 characters or which are not part of special cases. Affected parts: Wor_flow_hysteresis. Affected line: parameter Modelica.Units.SI.MassFlowRate mWor_flow_hysteresis = mWor_flow_min + (mWor_flow_max - mWor_flow_min) * 0.1 "Hysteresis for turning off the cycle when flow too low" annotation(Dialog(group="Cycle"));

7: Name 'on_actual' contains parts with more/less than 3 characters or which are not part of special cases. Affected parts: on_actual. Affected line: Modelica.Blocks.Interfaces.BooleanOutput on_actual "Actual on off status of the cycle" annotation (Placement(transformation( extent={{100,-20},{140,20}}), iconTransformation(extent={{100,-10},{ 120,10}})));

IBPSA\Fluid\CHPs\OrganicRankine\BaseClasses\FixedEvaporating.mo
1: Name 'dTPinEva_set' contains parts with more/less than 3 characters or which are not part of special cases. Affected parts: Eva_set. Affected line: parameter Modelica.Units.SI.TemperatureDifference dTPinEva_set( final min = 0) "Set evaporator pinch point temperature difference" annotation(Dialog(group="Evaporator"));

5: Name 'mWor_flow_hysteresis' contains parts with more/less than 3 characters or which are not part of special cases. Affected parts: Wor_flow_hysteresis. Affected line: parameter Modelica.Units.SI.MassFlowRate mWor_flow_hysteresis( final min = 0) "Hysteresis for turning off the cycle when flow too low" annotation(Dialog(group="Cycle"));

6: Name 'on_actual' contains parts with more/less than 3 characters or which are not part of special cases. Affected parts: on_actual. Affected line: Modelica.Blocks.Interfaces.BooleanOutput on_actual = ena and hys.y "Actual on off status of the cycle" annotation (Placement(transformation( extent={{100,-20},{140,20}}), iconTransformation(extent={{100,-10},{120, 10}})));
```

hcasperfu commented:
@FWuellhorst -
Thank you very much for the extensive review and terribly sorry for taking me so long to get back to this.
1. Much of the system and control assumptions of this model were based on [Quoilin et al. 2011](https://www.sciencedirect.com/science/article/abs/pii/S0306261911000183). It stated that "The most common control strategy is to define a constant evaporating temperature and superheating", although the evaporating temperature $T_{eva}$ needs to be determined by prior analysis. It then based the control strategy on the working fluid mass flow rate $\dot m_{wor}$ because it was the most "easily accessible measurement". The cycle reacts to a variable heat source by adjusting $\dot m_{wor}$ to maintain the pinch point at a fixed $T_{eva}$.
2. The validation model was primarily designed to test whether the model breaks at zero flow rates, as is how they are intended in the Buildings library. We do not have relevant data readily available for experiment-based tests.
3. Indeed the specific terms, interestingly, are not listed in the naming convention. But examples using similar names can be found in the Buildings library.

I will work on your inline comments shortly.

hcasperfu commented on IBPSA/Fluid/CHPs/OrganicRankine/Cycle.mo:
```diff
@@ -0,0 +1,494 @@
+can be configured to have first order dynamics.
+The fluid stream 1 (using <code>Medium1</code>, <code>port_a1</code>, etc.)
+is the evaporator hot fluid carrying waste heat
+and the stream 2 is the condenser cold fluid carrying the cooling fluid.
```
Modified as suggested in eb6a21c.

hcasperfu commented on IBPSA/Fluid/CHPs/OrganicRankine/Cycle.mo:
```diff
@@ -0,0 +1,494 @@
+The superheating temperature difference <i>&Delta;T<sub>sup</sub></i>
+is minimized, meaning it is zero whenever possible; otherwise it assumes
+the smallest value not to cause the expander outlet state to fall
+under the dome. Subcooling after the condenser is not considered.
```
Added clarification in eb6a21c. In grad school in the US it was always called the "dome", so it appears typical to me.

hcasperfu commented on IBPSA/Fluid/CHPs/OrganicRankine/Cycle.mo:
```diff
@@ -0,0 +1,494 @@
+<h4>Constraints</h4>
+<p>
+The ORC system controls <i>m&#775;<sub>w</sub></i> to maintain the prescribed
+evaporator PP temperature difference set point.
```
This abbreviation was defined uptext. If you feel strongly I am open to drop it.

FWuellhorst replied:
All fine

hcasperfu commented on IBPSA/Fluid/CHPs/OrganicRankine/Cycle.mo:
```diff
@@ -0,0 +1,494 @@
+evaporator PP temperature difference set point.
+Although the model does not implement this as a control loop,
+an upper limit and a lower limit are imposed on
+<i>m&#775;<sub>w</sub></i> to reflect the characteristics of a sized cycle.
```
Added clarification in eb6a21c.

hcasperfu commented on IBPSA/Fluid/CHPs/OrganicRankine/Cycle.mo:
```diff
@@ -0,0 +1,494 @@
+Specific enthalpy and specific entropy values are provided as support points
+on the saturated liquid line, the saturated vapor line, and
+a superheated vapor line (called the reference line).
+The values of these support points were obtained using CoolProp
+(<a href=\"https://www.coolprop.org\">https://www.coolprop.org</a>;
```
This is a good idea!
@mwetter - Can you please provide guidance on how I can include the python scripts that I wrote to generate working fluid data records?

hcasperfu replied:
Example script added 6bb50d7.

hcasperfu commented on IBPSA/Fluid/CHPs/OrganicRankine/Cycle.mo:
```diff
@@ -0,0 +1,494 @@
+the expander inlet or outlet depending on the shape of the cycle.
+The isobaric lines are not straight in this section, but they are
+assumed linear so that the method for <i>B</i> can be applied using the
+saturated vapor line and the reference line, albeit with less accuracy.
```
- This is discussed in detail in the conference paper. The paper reference will be included once it is published.
- Can you explain what you mean by "line 3"?

FWuellhorst replied:
I meant the reference line 
![image](https://github.com/user-attachments/assets/071f8e71-5e31-4bb4-84eb-3b3223f80ec9)

hcasperfu replied:
This line is set once the data record is generated. Linking back to CoolProp to re-generate the record, in my opinion, defeats the purpose of this model, as it is (in the conf paper) advertised to be self-contained.
A better way to construct this data record is also a point of discussion in the conference paper, which will be cited in the documentation once it has a DOI.

FWuellhorst replied:
Ok, sounds good!

hcasperfu commented on IBPSA/Fluid/CHPs/OrganicRankine/Cycle.mo:
```diff
@@ -0,0 +1,494 @@
+saturated vapor line and the reference line, albeit with less accuracy.
+</li>
+</ul>
+<p>
+The cycle can be completely defined by providing the following quantities:
```
I don't see why a new section needs to be started here. The following part is still talking about how the thermodynamic properties are computed.

hcasperfu commented on IBPSA/Fluid/CHPs/OrganicRankine/Validation/VariableSource.mo:
```diff
@@ -0,0 +1,155 @@
+within IBPSA.Fluid.CHPs.OrganicRankine.Validation;
+model VariableSource
```
I agree.
I have, however, grown accustomed to the fact that the validation models test whether things break when flow rates get in and out of zero.

FWuellhorst replied:
This is maybe a point for discussion in the working group.

hcasperfu commented on IBPSA/Fluid/CHPs/OrganicRankine/BaseClasses/Validation/DryFluid.mo:
```diff
@@ -0,0 +1,47 @@
+within IBPSA.Fluid.CHPs.OrganicRankine.BaseClasses.Validation;
+model DryFluid
+  "Organic Rankine cycle with a dry working fluid"
+  extends Modelica.Icons.Example;
```
This analysis will be in the conference paper to be published. In my opinion, testing each data record exhaustively will make the package more bloated than necessary. I know we do something similar for the functions in the math package, but I think that is very different.

FWuellhorst replied:
Makes sense!

hcasperfu commented on IBPSA/Fluid/CHPs/OrganicRankine/Data/Generic.mo:
```diff
@@ -0,0 +1,40 @@
+Record containing parameters for working fluid properties.
+A figure in the documentation of
+<a href=\"Modelica://IBPSA.Fluid.CHPs.OrganicRankine.BottomingCycle\">
+IBPSA.Fluid.CHPs.OrganicRankine.BottomingCycle</a>
```
Fixed in 0c97806.

hcasperfu commented on IBPSA/Fluid/CHPs/package.order:
```diff
@@ -0,0 +1 @@
+OrganicRankine
```
`CHPs` is the package hosting the ORC model in Buildings library. This package contains a topping cycle model in parallel with the ORC model.
I agree `ConstantEvaporationSteadyState` is a suitable name for this model, but as we do not currently have plan to expand this model further due to end of project, I am in favour of short and crisp names as they are now. I am still open to rename them if you feel strongly.

FWuellhorst replied:
I like short names, as well, just mentioned it based on the review you gave for the HP. And I thought it is a fair point to call the model based on what it models, as users often do not read documentation. I would propose `Fluid.CHPs.ORC.ConstantEvaporationORC` or `Fluid.CHPs.ORC.ConstantEvaporation` or `Fluid.CHPs.ORC.WasteWater`. Right now, the user has to understand the physics and know ORC to see that it is only for one type of application.

hcasperfu replied:
Point taken.
Class renamed to `ConstantEvaporation` in 6612f5a.

hcasperfu commented on IBPSA/Fluid/CHPs/OrganicRankine/Cycle.mo:
```diff
@@ -0,0 +1,494 @@
+    "Nominal pressure drop of the cold fluid in condenser"
+    annotation(Dialog(group="Condenser"));
+  parameter Modelica.Units.SI.TemperatureDifference dTPinCon(
+    final min = 0) = 10
```
This number is from [Quoilin et al. 2011](https://www.sciencedirect.com/science/article/abs/pii/S0306261911000183).

hcasperfu commented on IBPSA/Resources/Images/Fluid/CHPs/OrganicRankine/CycleArchitecture.png:
Can you please explain why this was not clear?
$\Delta T_{pin,con} = T_{w,con} - T_{c,con}$ and is marked by the double arrow between those two horizontal lines.

<img width="199" alt="image" src="https://github.com/user-attachments/assets/955b85ce-987f-4156-9ba0-66aae0a962a1">

FWuellhorst replied:
On the top, the pinch is correctly displayed on the line:
![image](https://github.com/user-attachments/assets/05f8939b-7d1e-40b5-9ea2-4ae9c6975ccd)
but here, it is moved slightly to the left. At first glance, it seemed too far away from the actual pinch. Maybe with using color, e.g. red, you could put the arrow above the actual pinch point. But the current version is fine as well

hcasperfu commented on IBPSA/Fluid/CHPs/OrganicRankine/Cycle.mo:
```diff
@@ -0,0 +1,494 @@
+    "Set true to specify working fluid evaporating pressure instead of temperature"
+    annotation(Dialog(group="Evaporator"));
+  parameter Modelica.Units.SI.ThermodynamicTemperature TWorEva =
+    max(pro.T)*2/3 + min(pro.T)*1/3
```
This is a placeholder for handling "conditional" parameters. The intention is that the user can either use $T_{wor,eva}$ or $p_{wor,eva}$ to specify the cycle. Even though only one of these two parameters is needed, Modelica would complain unless both are given a value. Hence a hopefully reasonable default value is given here.

FWuellhorst replied:
Makes sense. Can you document the reason for this default if users wonder? Because it looks like a physical equation, even though it is just an educated guess.

hcasperfu replied:
Updated documentation in a7934de.

hcasperfu commented on IBPSA/Fluid/CHPs/OrganicRankine/Cycle.mo:
```diff
@@ -0,0 +1,494 @@
+where the subscripts are
+<i>eva</i> for evaporator,
+<i>exp</i> for expander,
+<i>h</i> for hot fluid of the evaporator, i.e. the fluid carrying heat,
+<i>pum</i> for pump, and
+<i>w</i> for working fluid.
```
I agree. However, as the final conference paper is accepted, I am inclined to keep the figures intact because they are used in the same places. I can revise them if we get a chance to further submit a final version, or when we expand it to a journal paper, in which case the paper reference will be updated.

FWuellhorst replied:
Sounds good. I found them confusing only at first glance

hcasperfu commented on IBPSA/Fluid/CHPs/OrganicRankine/Cycle.mo:
```diff
@@ -0,0 +1,494 @@
+    "Pump efficiency"
+    annotation(Dialog(group="Cycle"));
+
+  Modelica.Blocks.Interfaces.BooleanInput ena
```
`ena` is used in the Buildings library and, if memory serves, was adopted here in favour of `on` during Buildings review.
In my opinion, `on` is not clear whether it is input or output and `u` is not very descriptive, whereas `ena` expresses the action of enabling the equipment more clearly. In more context, at Buildings we have been trying to use more descriptive variable names, for example, instead of `port_a` we call it `port_aWatIn` whenever possible.

FWuellhorst replied:
Makes sense. As the cycle may shut off even if ena is True, the name is also fitting.

hcasperfu commented on IBPSA/Fluid/CHPs/OrganicRankine/Validation/VariableSource.mo:
```diff
@@ -0,0 +1,155 @@
+  connect(THotIn_set.y, souHot.T_in) annotation (Line(points={{-59,10},{-50,10},
+          {-50,34},{-42,34}}, color={0,0,127}));
+  annotation(experiment(StopTime=300,Tolerance=1E-6),
+  __Dymola_Commands(file="modelica://IBPSA/Resources/Scripts/Dymola/Fluid/CHPs/OrganicRankine/Validation/VariableSource.mos"
+  "Simulate and plot"),
+  Documentation(info="<html>
```
I see your point.
In my opinion, the user should be able to see that there is no working fluid flow rate, the cycle is off, and there is no heat transfer. This is well explained and documented in the package and this model is intended to demonstrate this behaviour.

FWuellhorst commented:
@hcasperfu 
Sure, glad to provide feedback.
Regarding your other points:
1. Ok, but this paper is not mentioned in the doc. Maybe I would state it in the general information "...  as a bottoming cycle based on XY et al.". Based on your clarifications, this model is only suited for waster-water applications, or? We have a project for Carnot-Batteries. In this case, the constant evaporation temperature is of no good and the model would fail to work. This is why I feel strongly about a fitting name instead of just ORC.
2. To be discussed in working group. 
3. Ok, maybe you have to update the naming convention wiki then.

FWuellhorst commented on IBPSA/Resources/Images/Fluid/CHPs/OrganicRankine/CycleArchitecture.png:
These files must be renamed to ConstantEvaporation, as they were renamed in the doc, I guess with the refactor script. Maybe this is a small bug in buildingspy when changing names of file references but the files are not renamed?

hcasperfu replied:
The names of the image files are not related to the name of the model. So keeping them unchanged is intended.

FWuellhorst reviewed (approved):
@hcasperfu All changes look good, only the images have to be renamed, as well. I don't think adding the script directly is necessary, you could do that upon adding the reference to the conference paper in my opinion. I will be on vacation in the upcoming weeks and not able to further review this PR if you do further changes.

hcasperfu commented:
@mwetter -
This PR has cleared external review and ready to merge.

mwetter commented on IBPSA/Resources/Python-Sources/MakeORCFluidRecord.py:
```diff
@@ -0,0 +1,110 @@
+       T = np.linspace(TMin,TMax,10)
+       p = CP.PropsSI('P','T',T,'Q',1,nCP)
+       rhoLiq = CP.PropsSI('D','T',T,'Q',0,nCP)
+       pMin = CP.PropsSI('P','T',TMin,'Q',1,nCP)
+       pMax = CP.PropsSI('P','T',TMax,'Q',1,nCP)
```
`pMin` and `pMax` are not used.

hcasperfu replied:
Deleted in 46a7301.

mwetter reviewed (changes requested):
This result is wrong:
![image](https://github.com/user-attachments/assets/e5972709-7e8d-49e1-a45c-a8a41a259eca)
Produced from 
`
simulateModel("IBPSA.Fluid.CHPs.OrganicRankine.Examples.ORCHotWater",
    method="Cvode", stopTime=900, tolerance=1e-06, resultFile="ORCHotWater");
`
Also, generating the html documentation gives
`
Can not resolve link "Modelica://IBPSA.DHC.ETS.Combined.Subsystems.Validation.Chiller" in [IBPSA.Fluid.CHPs.OrganicRankine.Examples.ORCHotWater](modelica://IBPSA.Fluid.CHPs.OrganicRankine.Examples.ORCHotWater)
`

In Dymola, `IBPSA.Fluid.CHPs.OrganicRankine.Examples.ORCHotWater` gives warning due to non-initialized iteration variables. Set a start value to avoid this.

hcasperfu commented:
@mwetter -

Thanks for the review!
1. The incorrect variable is now deleted in 40b0f48. I messed up the sign in the equation which also had a guard against zero. Because this top-level variable was not used anywhere and the same variable is available in the component, I simply deleted it. The same variable at a lower level was correct.
<img width="515" alt="image" src="https://github.com/user-attachments/assets/3839d669-74f5-47b4-85b8-24f1904bef03">

2. I removed the unresolved link in 2805b60. It referred to a model only available in Buildings and it did was say "this is the same as that".
3. I added start values to suppress those Dymola translation warnings in 073ad76.
